### PR TITLE
Module prepend

### PR DESCRIPTION
**Pull request by archseer**
Things left to port:
- [x] https://github.com/ruby/ruby/commit/8ddbbb33242f413845079b0274c28331cb4aa0f5
- [x] https://github.com/ruby/ruby/commit/f113ab552c7f83bd12f4136bce7fb8ee89819adb
- [x] https://github.com/ruby/ruby/commit/5922c954614e5947a548780bb3b894626affe6dd
- [X] https://github.com/ruby/ruby/commit/59270c607bb4dd053bb4b6ecc0719de011a0eb55
- [X] https://github.com/ruby/ruby/commit/edb1fc4eb2f9b0bd69398362ae40100adfd5577e
- [x] https://github.com/ruby/ruby/commit/a58c224dcff50eec3149df0f91b8a0f2e8099db6
- [ ] https://github.com/ruby/ruby/search?utf8=%E2%9C%93&q=RCLASS_ORIGIN

### CODE REVIEW

File: src/class.c [hunk 1]
```diff
@@ -759,47 +769,122 @@ boot_defclass(mrb_state *mrb, struct RClass *super)
...
-    m = m->super;
+    module = module->super;
+  }
+  return 0;
+}
+
+MRB_API void
+mrb_include_module(mrb_state *mrb, struct RClass *c, struct RClass *m)
+{
+  include_module_at(mrb, c, c->origin, m, 1);
```
Comment (archseer) [inline src/class.c]: Include must also check `changed` now.

**Comment by IceDragon200** [comment]
@archSeer: Copied over the tests, some of them fail though, give it a run

**Comment by archseer** [comment]
```
Fail: test_prepend_instance_methods_false
 - Assertion[1] Failed: [ruby-dev:45863]
    Expected: [:m1]
      Actual: []
 - Assertion[2] Failed: [ruby-dev:45863]
    Expected: [:m1]
      Actual: []
```

This one requires [this patch](https://bugs.ruby-lang.org/issues/6660):

``` diff
diff --git a/class.c b/class.c
index a642ff0..9d346ee 100644
--- a/class.c
+++ b/class.c
@@ -927,7 +927,7 @@ static VALUE
 class_instance_method_list(int argc, VALUE *argv, VALUE mod, int obj, int (*func) (st_data_t, st_data_t, st_data_t))
 {
     VALUE ary;
-    int recur;
+    int recur, prepended = 0;
     st_table *list;

     if (argc == 0) {
@@ -939,10 +939,15 @@ class_instance_method_list(int argc, VALUE *argv, VALUE mod, int obj, int (*func
    recur = RTEST(r);
     }

+    if (!recur && RCLASS_ORIGIN(mod) != mod) {
+   mod = RCLASS_ORIGIN(mod);
+   prepended = 1;
+    }
+
     list = st_init_numtable();
     for (; mod; mod = RCLASS_SUPER(mod)) {
    if (RCLASS_M_TBL(mod)) st_foreach(RCLASS_M_TBL(mod), method_entry_i, (st_data_t)list);
-   if (BUILTIN_TYPE(mod) == T_ICLASS) continue;
+   if (BUILTIN_TYPE(mod) == T_ICLASS && !prepended) continue;
    if (obj && FL_TEST(mod, FL_SINGLETON)) continue;
    if (!recur) break;
     }
```

**Comment by archseer** [comment]
Applied patch.

**Comment by archseer** [comment]
The segfault one also seems to run fine for me, `test_prepend_module_ancestors`.

**Comment by archseer** [comment]
There's a whole bunch of code changed inside `vm_method.c`, which we haven't ported: https://github.com/ruby/ruby/commit/8ddbbb33242f413845079b0274c28331cb4aa0f5#diff-6cea7716df08f2b7fdc5fb30c7963884L169

**Comment by archseer** [comment]
@IceDragon200 Also rebased on latest master.

**Comment by IceDragon200** [comment]
@archSeer mruby doesn't seem to have redefine guards, so we won't need the `vm_method.c` changes, the `search_method` is done in `class.c` using `mrb_method_search_vm`, which already has a null check for method tables.

File: src/class.c [hunk 2]
```diff
@@ -767,47 +771,132 @@ boot_defclass(mrb_state *mrb, struct RClass *super)
   return c;
 }
 
-MRB_API void
-mrb_include_module(mrb_state *mrb, struct RClass *c, struct RClass *m)
+static void
+boot_initmod(mrb_state *mrb, struct RClass *mod)
```
Comment (archseer) [inline src/class.c:775]: I think this should either be removed, or renamed and extended to function like `class_alloc` in ruby.

File: src/class.c [hunk 3]
```diff
@@ -767,47 +771,132 @@ boot_defclass(mrb_state *mrb, struct RClass *super)
...
-  struct RClass *ins_pos;
+  mod->origin = mod;
+  mod->mt = kh_init(mt, mrb);
+}
 
-  ins_pos = c;
-  while (m) {
-    struct RClass *p = c, *ic;
+static struct RClass*
+include_class_new(mrb_state *mrb, struct RClass *m, struct RClass *super)
```
Comment (archseer) [inline src/class.c:782]: This method is used exactly once.

File: src/class.c [hunk 4]
```diff
@@ -767,47 +771,132 @@ boot_defclass(mrb_state *mrb, struct RClass *super)
...
+  if (m->tt == MRB_TT_ICLASS) {
+    ic->c = m->c;
+  } else {
+    ic->c = m;
+  }
+  return ic;
+}
+
+MRB_API int
+include_module_at(mrb_state *mrb, struct RClass *klass, struct RClass *c, struct RClass *module, int search_super)
```
Comment (archseer) [inline src/class.c]: We might want to rename these variables to what they were before in mruby. For instance I thought that my way of doing `include_module_at(mrb_state *mrb, struct RClass *c, struct RClass *ins_pos, struct RClass *m, int search_super)` made much more sense, `c` being class, `ins_pos`/`cur` being the position where modules are inserted and `m` being the module that is being included.

File: src/class.c (hunk 2 above)
Comment (IceDragon200) [inline src/class.c:775]: I'm bad a naming, I tried to make it match the existing `boot_defclass` which is for bootstrapping the first 4 classes (BasicObject, Object, Class and Module). `bootstrap_initialize_module` is what its read as, but yeah, feel free to rename it, its not a class_alloc, since it takes a RClass module and initializes its fields

File: src/class.c (hunk 3 above)
Comment (IceDragon200) [inline src/class.c:782]: It moves the noisy `ICLASS` creation into a function, this code is very distracting when placed into  `include_module_at`, since its already a complicated function, try describing the function without using too many `and` and you'll see what I mean.

tl;dr separation of concerns

File: src/class.c (hunk 2 above)
Comment (IceDragon200) [inline src/class.c:775]: Better yet, we should find all occurences of module creation and run it through one function that does proper initialization, like class
